File: pa6-cky-v3/java/src/nlpclass/ling/Trees.java

package nlpclass.ling;

import nlpclass.util.Filter;

import java.io.IOException;
import java.io.PushbackReader;
import java.io.Reader;
import java.io.StringReader;
import java.util.*;

/**
 * Tools for displaying, reading, and modifying trees.
 *
 * @author Dan Klein
 * @author Paul Baumstarck (added GENIA and BioIE readers 2008.4)
 */
public class Trees {

    public static interface TreeTransformer<E> {
        Tree<E> transformTree(Tree<E> tree);
    }

    public static class FunctionNodeStripper implements TreeTransformer<String> {
        public Tree<String> transformTree(Tree<String> tree) {
            String transformedLabel = tree.getLabel();
            int cutIndex = transformedLabel.indexOf('-');
            int cutIndex2 = transformedLabel.indexOf('=');
            if (cutIndex2 > 0 && (cutIndex2 < cutIndex || cutIndex == -1))
                cutIndex = cutIndex2;
            cutIndex2 = transformedLabel.indexOf('^');
            if (cutIndex2 > 0 && (cutIndex2 < cutIndex || cutIndex == -1))
                cutIndex = cutIndex2;
            cutIndex2 = transformedLabel.indexOf(':');
            if (cutIndex2 > 0 && (cutIndex2 < cutIndex || cutIndex == -1))
                cutIndex = cutIndex2;

            if (cutIndex > 0 && !tree.isLeaf()) {
                transformedLabel = new String(transformedLabel.substring(0,
                        cutIndex));
            }
            if (tree.isLeaf()) {
                return new Tree<String>(transformedLabel);
            }
            List<Tree<String>> transformedChildren = new ArrayList<Tree<String>>();
            for (Tree<String> child : tree.getChildren()) {
                transformedChildren.add(transformTree(child));
            }
            return new Tree<String>(transformedLabel, transformedChildren);
        }
    }

    public static class EmptyNodeStripper implements TreeTransformer<String> {
        public Tree<String> transformTree(Tree<String> tree) {
            String label = tree.getLabel();
            if (label.equals("-NONE-")) {
                return null;
            }
            if (tree.isLeaf()) {
                return new Tree<String>(label);
            }
            List<Tree<String>> children = tree.getChildren();
            List<Tree<String>> transformedChildren = new ArrayList<Tree<String>>();
            for (Tree<String> child : children) {
                Tree<String> transformedChild = transformTree(child);
                if (transformedChild != null)
                    transformedChildren.add(transformedChild);
            }
            if (transformedChildren.size() == 0)
                return null;
            return new Tree<String>(label, transformedChildren);
        }
    }

    public static class XOverXRemover<E> implements TreeTransformer<E> {
        public Tree<E> transformTree(Tree<E> tree) {
            E label = tree.getLabel();
            List<Tree<E>> children = tree.getChildren();
            while (children.size() == 1 && !children.get(0).isLeaf()
                    && label.equals(children.get(0).getLabel())) {
                children = children.get(0).getChildren();
            }
            List<Tree<E>> transformedChildren = new ArrayList<Tree<E>>();
            for (Tree<E> child : children) {
                transformedChildren.add(transformTree(child));
            }
            return new Tree<E>(label, transformedChildren);
        }
    }

    public static class StandardTreeNormalizer implements
            TreeTransformer<String> {
        EmptyNodeStripper emptyNodeStripper = new EmptyNodeStripper();
        XOverXRemover<String> xOverXRemover = new XOverXRemover<String>();
        FunctionNodeStripper functionNodeStripper = new FunctionNodeStripper();

        public Tree<String> transformTree(Tree<String> tree) {
            tree = functionNodeStripper.transformTree(tree);
            tree = emptyNodeStripper.transformTree(tree);
            tree = xOverXRemover.transformTree(tree);
            return tree;
        }
    }

    public static class TreeReader {
        public static String ROOT_LABEL = "ROOT";
        
        PushbackReader in;
        Tree<String> nextTree;

        public boolean hasNext() {
            return (nextTree != null);
        }

        public Tree<String> next() {
            if (!hasNext())
                throw new NoSuchElementException();
            Tree<String> tree = nextTree;
            nextTree = readRootTree();
            return tree;
        }
        
        public Tree<String> readRootTree() {
            throw new RuntimeException("readRootTree() undefined.");
        }

        public int peek() throws IOException {
            int ch = in.read();
            in.unread(ch);
            return ch;
        }

        public String readLabel() throws IOException {
            readWhiteSpace();
            return readText();
        }
        
        public Tree<String> readLeaf() throws IOException {
            String label = readText();
            return new Tree<String>(label);
        }
        
        public String readText() throws IOException {
            StringBuilder sb = new StringBuilder();
            int ch = in.read();
            while (!isWhiteSpace(ch) && !isLeftParen(ch) && !isRightParen(ch)) {
                sb.append((char) ch);
                ch = in.read();
            }
            in.unread(ch);
            //      System.out.println("Read text: ["+sb+"]");
            return sb.toString().intern();
        }
        
        public void readLeftParen() throws IOException {
            //      System.out.println("Read left.");
            readWhiteSpace();
            int ch = in.read();
            if (!isLeftParen(ch)) {
                throw new RuntimeException("Format error reading tree.");
            }
        }

        public void readRightParen() throws IOException {
            //      System.out.println("Read right.");
            readWhiteSpace();
            int ch = in.read();
            if (!isRightParen(ch)) {
                throw new RuntimeException("Format error reading tree.");
            }
        }

        public void readWhiteSpace() throws IOException {
            int ch = in.read();
            while (isWhiteSpace(ch)) {
                ch = in.read();
            }
            in.unread(ch);
        }

        public boolean isWhiteSpace(int ch) {
            return (ch == ' ' || ch == '\t' || ch == '\f' || ch == '\r' || ch == '\n');
        }

        public boolean isLeftParen(int ch) {
            return ch == '(';
        }

        public boolean isRightParen(int ch) {
            return ch == ')';
        }
        
        public boolean isSemicolon(int ch) {
            return ch == ';';
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }
    }

    public static class BioIETreeReader extends TreeReader implements Iterator<Tree<String>> {
        
        public Tree<String> readRootTree() {
            try {
                for (;;) {
                    readCommentsAndWhiteSpace();
                    if (!isLeftParen(peek()))
                        return null;
                    in.read(); // (
                    String str = readText();
                    if ( str.equals("SENT") )
                        break; // read the sentence
                    else if ( str.equals("SEC") ) {
                        // read in the section, "("-less tree, then throw it away
                        readTree(false);
                    } else {
                        //System.out.println("Failed: "+str);
                        return null; // reading error
                    }
                }
                // We have stripped off "(SENT", so read a "("-less tree.
                return new Tree<String>(ROOT_LABEL, Collections.singletonList(readTree(false)));
            } catch (IOException e) {
                throw new RuntimeException("Error reading tree."+e.toString());
            }
        }

        private Tree<String> readTree(boolean matchparen) throws IOException {
            if ( matchparen )
                readLeftParen();
            String label = readColonizedLabel();
            //System.out.println("  "+label);
            List<Tree<String>> children = readChildren();
            readRightParen();
            return new Tree<String>(label, children);
        }
        
        public String readColonizedLabel() throws IOException {
            readWhiteSpace();
            String ret = readText();
            int i = ret.indexOf(":");
            if ( i == -1 )
                return ret;
            else
                return ret.substring(0,i);
        }
        
        private List<Tree<String>> readChildren() throws IOException {
            readWhiteSpace();
            if (!isLeftParen(peek()))
                return Collections.singletonList(readLeaf());
            else
                return readChildList();
        }

        private List<Tree<String>> readChildList() throws IOException {
            List<Tree<String>> children = new ArrayList<Tree<String>>();
            readWhiteSpace();
            while (!isRightParen(peek())) {
                children.add(readTree(true));
                readWhiteSpace();
            }
            return children;
        }
                
        private void readCommentsAndWhiteSpace() throws IOException {
            int ch;
            for (;;) {
                readWhiteSpace();
                if ( !isSemicolon(peek()) )
                    return;
                // read a line
                ch = in.read();
                while ( ch != '\n' )
                    ch = in.read();
            }
        }

        public BioIETreeReader(Reader in) {
            this.in = new PushbackReader(in);
            nextTree = readRootTree();
        }
    }

    public static class PennTreeReader extends TreeReader implements Iterator<Tree<String>> {
        
        public Tree<String> readRootTree() {
            try {
                readWhiteSpace();
                if (!isLeftParen(peek()))
                    return null;
                return readTree(true);
            } catch (IOException e) {
                throw new RuntimeException("Error reading tree.");
            }
        }

        private Tree<String> readTree(boolean isRoot) throws IOException {
            readLeftParen();
            String label = readLabel();
            if (label.length() == 0 && isRoot)
                label = ROOT_LABEL;
            List<Tree<String>> children = readChildren();
            readRightParen();
            return new Tree<String>(label, children);
        }

        private List<Tree<String>> readChildren() throws IOException {
            readWhiteSpace();
            if (!isLeftParen(peek()))
                return Collections.singletonList(readLeaf());
            return readChildList();
        }

        private List<Tree<String>> readChildList() throws IOException {
            List<Tree<String>> children = new ArrayList<Tree<String>>();
            readWhiteSpace();
            while (!isRightParen(peek())) {
                children.add(readTree(false));
                readWhiteSpace();
            }
            return children;
        }

        public PennTreeReader(Reader in) {
            this.in = new PushbackReader(in);
            nextTree = readRootTree();
        }
    }

    public static class GENIATreeReader extends TreeReader implements Iterator<Tree<String>> {

        public Tree<String> readRootTree() {
            try {
                readWhiteSpace();
                if (!isLeftParen(peek()))
                    return null;
                return new Tree<String>(ROOT_LABEL, Collections.singletonList(readTree(false)));
            } catch (IOException e) {
                throw new RuntimeException("Error reading tree.");
            }
        }

        private Tree<String> readTree(boolean isRoot) throws IOException {
            readLeftParen();
            String label = readLabel();
            if (label.length() == 0 && isRoot)
                label = ROOT_LABEL;
            List<Tree<String>> children = readChildren();
            readRightParen();
            return new Tree<String>(label, children);
        }

        private List<Tree<String>> readChildren() throws IOException {
            List<Tree<String>> children = new ArrayList<Tree<String>>();
            readWhiteSpace();
            while (!isRightParen(peek())) {
                if ( isLeftParen(peek()) ) {
                    children.add(readTree(false));
                } else {
                    Tree<String> ret = readSlashLabel();
                    if ( ret != null )
                        children.add(ret);
                }
                readWhiteSpace();
            }
            return children;
        }
        
        private Tree<String> readSlashLabel() throws IOException {
            String label = readText();
            int i = label.lastIndexOf("/");
            if ( i == -1 ) return null;
            while ( i > 0 && label.charAt(i-1) == '\\' ) {
                i = label.lastIndexOf("/",i-1);
            }
            return new Tree<String>(label.substring(i+1),
                    Collections.singletonList(new Tree<String>( label.substring(0,i).replaceAll("\\\\\\/","\\/") )));
        }

        public GENIATreeReader(Reader in) {
            this.in = new PushbackReader(in);
            nextTree = readRootTree();
        }
    }

    /**
     * Renderer for pretty-printing trees according to the Penn Treebank indenting
     * guidelines (mutliline).  Adapted from code originally written by Dan Klein
     * and modified by Chris Manning.
     */
    public static class PennTreeRenderer {

        /**

File: pa6-cky-v3/python/ling/Trees.py
from Tree import Tree

# TODO: should I replace rendering of tree.label to str(tree.label)??

##################
# Class Methods
##################
ROOT_LABEL = "ROOT"

class TreeTransformer:
    """
        Abstract base class for different Tree transformation classes.
    """
    @classmethod
    def transform_tree(cls, tree):
        raise NotImplementedError()

class FunctionNodeStripper(TreeTransformer):

    @classmethod
    def transform_tree(cls, tree):
        transformed_label = tree.label
        cut_idx = transformed_label.find('-')
        cut_idx2 = transformed_label.find('=')
        if cut_idx2 > 0 and (cut_idx2 < cut_idx or cut_idx == -1):
            cut_idx = cut_idx2
        cut_idx2 = transformed_label.find('^')
        if cut_idx2 > 0 and (cut_idx2 < cut_idx or cut_idx == -1):
            cut_idx = cut_idx2
        cut_idx2 = transformed_label.find(':')
        if cut_idx2 > 0 and (cut_idx2 < cut_idx or cut_idx == -1):
            cut_idx = cut_idx2

        if cut_idx > 0 and not tree.is_leaf():
            transformed_label = transformed_label[:cut_idx]
        if tree.is_leaf():
            return Tree(transformed_label)

        transformed_children = []
        for child in tree.children:
            transformed_children.append(FunctionNodeStripper.transform_tree(child))

        return Tree(transformed_label, transformed_children)

class EmptyNodeStripper(TreeTransformer):

    @classmethod
    def transform_tree(cls, tree):
        label = tree.label
        if label == "-NONE-":
            return None
        if tree.is_leaf():
            return Tree(label)
        children = tree.children
        transformed_children = []
        for child in children:
            transformed_child = EmptyNodeStripper.transform_tree(child)
            if transformed_child is not None:
                transformed_children.append(transformed_child)
        if len(transformed_children) == 0:
            return None
        return Tree(label, transformed_children)

class XOverXRemover(TreeTransformer):

    @classmethod
    def transform_tree(cls, tree):
        label = tree.label
        children = tree.children
        while len(children) == 1 and not children[0].is_leaf() \
                and label == children[0].label:
            children = children[0].children
        transformed_children = []
        for child in children:
            transformed_children.append(XOverXRemover.transform_tree(child))
        return Tree(label, transformed_children)

class StandardTreeNormalizer(TreeTransformer):

    @classmethod
    def transform_tree(cls, tree):
        tree = FunctionNodeStripper.transform_tree(tree)
        tree = EmptyNodeStripper.transform_tree(tree)
        tree = XOverXRemover.transform_tree(tree)
        return tree

class TreeReader:
    """
        Abstract base class for tree readers.
        NOTE: Does not implement read_root_tree()
        NOTE: self.ff is an open file object for reading a file
    """

    def __iter__(self):
        return self

    def next(self):
        if self.next_tree is None:
            raise StopIteration
        else:
            tree = self.next_tree
            self.next_tree = self.read_root_tree()
            return tree

    # Java version of iterable...
    """
    def has_next(self):
        return self.next_tree is not None

    def next(self):
        if not self.has_next():
            raise LookupError("No more trees!")
        tree = self.next_tree
        self.next_tree = self.read_root_tree()
        return tree
    """

    def read_root_tree(self):
        raise NotImplementedError()

    def peek(self):
        ch = self.ff.read(1)  # read a byte
        self.ff.seek(-1, 1)   # move back one byte
        return ch

    def read_label(self):
        self.read_whitespace()
        return self.read_text()

    def read_leaf(self):
        label = self.read_text()
        return Tree(label)

    def read_text(self):
        s = []
        ch = self.ff.read(1)
        while not TreeReader.is_whitespace(ch) and \
                not TreeReader.is_left_paren(ch) and \
                not TreeReader.is_right_paren(ch):
            s.append(ch)
            ch = self.ff.read(1)
        self.ff.seek(-1, 1)
        return ''.join(s)

    def read_left_paren(self):
        self.read_whitespace()
        ch = self.ff.read(1)
        if not TreeReader.is_left_paren(ch):
            raise ValueError("Format error reading tree. Character %d." % \
                    (self.ff.tell() - 1))

    def read_right_paren(self):
        self.read_whitespace()
        ch = self.ff.read(1)
        if not TreeReader.is_right_paren(ch):
            import ipdb; ipdb.set_trace()
            raise ValueError("Format error reading tree. (filename: %s)" % self.ff.name)

    def read_whitespace(self):
        ch = self.ff.read(1)
        while TreeReader.is_whitespace(ch):
            ch = self.ff.read(1)
        self.ff.seek(-1, 1)

    @classmethod
    def is_whitespace(cls, ch):
        return ch == ' ' or ch == '\t' or ch == '\f' or ch == '\r' or ch == '\n'

    @classmethod
    def is_left_paren(cls, ch):
        return ch == '('

    @classmethod
    def is_right_paren(cls, ch):
        return ch == ')'

    @classmethod
    def is_semicolon(cls, ch):
        return ch == ';'

    def remove(self):
        return NotImplementedError()

class BioIETreeReader(TreeReader):

    def __init__(self, ff):
        self.ff = ff
        self.next_tree = self.read_root_tree()

    def read_root_tree(self):
        try:
            while True:
                self.read_comments_and_whitespace()
                if not TreeReader.is_left_paren(self.peek()):
                    return None
                self.ff.read(1)
                string = self.read_text()
                if string == "SENT":
                    break
                elif string == "SEC":
                    self.read_tree(False)
                else:
                    return None
            # Collections.singletonList(readTree(false)) ??
            return Tree(ROOT_LABEL, [self.read_tree(False)])
        except IOError:
            raise Exception("Error reading tree: %s\n" % self.ff.name)

    def read_tree(self, matchparen):
        if matchparen:
            self.read_left_paren()
        label = self.read_colonized_label()
        children = self.read_children()
        self.read_right_paren()
        return Tree(label, children)

    def read_colonized_label(self):
        self.read_whitespace()
        ret = self.read_text()
        i = ret.find(':')
        if i == -1:
            return ret
        else:
            return ret[:i]

    def read_children(self):
        self.read_whitespace()
        if not TreeReader.is_left_paren(self.peek()):
            return [self.read_leaf()] # Collections.singletonList(readLeaf())
        else:
            return self.read_child_list()

    def read_child_list(self):
        children = []
        self.read_whitespace()
        while not TreeReader.is_right_paren(self.peek()):
            children.append(self.read_tree(True))
            self.read_whitespace()
        return children

    def read_comments_and_whitespace(self):
        while True:
            self.read_whitespace()
            if not TreeReader.is_semicolon(self.peek()):
                return
            ch = self.ff.read(1)
            while not ch == '\n':
                ch = self.ff.read(1)

class PennTreeReader(TreeReader):

    def __init__(self, ff):
        self.ff = ff
        self.next_tree = self.read_root_tree()

    def read_root_tree(self):
        try:
            self.read_whitespace()
            if not TreeReader.is_left_paren(self.peek()):
                return None
            return self.read_tree(True)
        except IOError:
            raise Exception('Error reading tree: %s\n' % self.ff.name)

    def read_tree(self, is_root):
        self.read_left_paren()
        label = self.read_label()
        if len(label) == 0 and is_root:
            label = ROOT_LABEL
        children = self.read_children()
        self.read_right_paren()
        return Tree(label, children)

    def read_children(self):
        self.read_whitespace()
        if not TreeReader.is_left_paren(self.peek()):
            return [self.read_leaf()] # Collections.singletonList
        return self.read_child_list()

    def read_child_list(self):
        children = []
        self.read_whitespace()
        while not TreeReader.is_right_paren(self.peek()):
            children.append(self.read_tree(False))
            self.read_whitespace()
        return children

class GENIATreeReader(TreeReader):

    def __init__(self, ff):
        self.ff = ff
        self.next_tree = self.read_root_tree()

    def read_root_tree(self):
        try:
            self.read_whitespace()
            if not TreeReader.is_left_paren(self.peek()):
                return None
            return Tree(ROOT_LABEL, [self.read_tree(False)])
        except IOError:
            raise Exception("Error reading tree: %s\n" % self.ff.name)

    def read_tree(self, is_root):
        self.read_left_paren()
        label = self.read_label()
        if len(label) == 0 and is_root:
            label = ROOT_LABEL
        children = self.read_children()
        self.read_right_paren()
        return Tree(label, children)

    def read_children(self):
        children = []
        self.read_whitespace()
        while not TreeReader.is_right_paren(self.peek()):
            if TreeReader.is_left_paren(self.peek()):
                children.append(self.read_tree(False))
            else:
                ret = self.read_slash_label()
                if ret is not None:
                    children.append(ret)
            self.read_whitespace()
        return children

    def read_slash_label(self):
        label = self.read_text()
        i = label.rfind('/')
        if i == -1:
            return None
        while i > 0 and label[i-1] == '\\':
            i = label.rfind('/', 0, i-1)
        child_label = label[:i].replace('\\\\\\/', '\\/')
        return Tree(label[i+1:], [Tree(child_label)])

class PennTreeRenderer:
    """
        Renderer for pretty-printing trees according to the Penn Treebank indenting
        guidelines (mutliline).  Adapted from code originally written by Dan Klein
        and modified by Chris Manning.
    """
    @classmethod
    def render(cls, tree):
        s = []
        PennTreeRenderer.render_tree(tree, 0, False, False, False, True, s)
        s.append('\n')
        return ''.join(s)

    @classmethod
    def render_tree(cls, tree, indent, parent_label_null, first_sibling, \
            left_sibling_preterminal, top_level, s):
        # Condition for staying on same line in Penn Treebank
        suppress_indent = parent_label_null or \
                (first_sibling and tree.is_preterminal()) or \
 